lib/origen_testers/smartest_based_tester/base/processors/flag_optimizer.rb

```diff
@@ -0,0 +1,107 @@
+module OrigenTesters
+  module SmartestBasedTester
+    class Base
+      module Processors
+        class FlagOptimizer < ATP::Processor
+          # This optimizes the AST such that any adjacent flow noes that
+          #
+          # For example this AST:
+          #   (flow
+          #     (test
+          #       (name "test1")
+          #       (on-fail
+          #         (set-run-flag "t1_FAILED")))
+          #     (run-flag "t1_FAILED" true
+          #       (test
+          #         (name "test2"))
+          #       (test
+          #         (name "test3")))
+          #     (test
+          #       (name "test4")
+          #       (on-pass
+          #         (set-run-flag "t4_PASSED"))
+          #       (on-fail
+          #         (continue)))
+          #     (run-flag "t4_PASSED" true
+          #       (test
+          #         (name "test5"))
+          #       (test
+          #         (name "test6"))))
+          #
+          # Will be optimized to this:
+          #   (flow
+          #     (test
+          #       (name "test1")
+          #       (on-fail
+          #         (test
+          #           (name "test2"))
+          #         (test
+          #           (name "test3"))))
+          #     (test
+          #       (name "test4")
+          #       (on-pass
+          #         (test
+          #           (name "test5"))
+          #         (test
+          #           (name "test6")))
+          #       (on-fail
+          #         (continue))))
+
+          # Only run this on top level flow and consider adjacent nodes, no need for
+          # looking at nested conditions.
+          def on_flow(node)
+            name, *nodes = *node
+            results = []
+            node_a = nil
+            nodes.each do |node_b|
+              if node_a && node_a.type == :test && node_b.type == :run_flag
+                result, node_a = on_boolean_substitution(node_a, node_b)
+                results << result
+              else
+                results << node_a unless node_a.nil?
+                node_a = node_b
+              end
+            end
+            results << node_a unless node_a.nil?
+            node.updated(nil, [name] + results)
+          end
+
+          # Given two adjacent nodes, where the first (a) is a test and the second (b)
+          # is a run_flag, determine if (a) conditionally sets the same flag that (b)
+          # uses.  If it does, do a logical replacement, if not, move on quietly.
```

Comment:
I think the name of this should be changed, on_boolean_substitution doesn't make sense in this context.
Should be called remove_run_flag or similar.